core/jobs/batch_jobs/opportunity_management_jobs.py

```diff
@@ -43,19 +45,192 @@
+                opportunity_models.SkillOpportunityModel.get_all(
+                    include_deleted=False))
+        )
+
+        unused_delete_result = (
+            skill_opportunity_model
+            | beam.Map(lambda model: model.key)
+            | 'Delete all models' >> ndb_io.DeleteModels()
+        )
+
+        return (
+            skill_opportunity_model
+            | 'Create job run result' >> (
+                job_result_transforms.CountObjectsToJobRunResult())
+        )
+
+
+class GenerateSkillOpportunityModelJob(base_jobs.JobBase):
+    """Job for regenerating SkillOpportunityModel.
+
+    NOTE: The DeleteSkillOpportunityModelJob must be run before this
+    job.
+    """
+
+    @staticmethod
+    def _count_unique_question_ids(
+        question_skill_link_models: List[
+            question_models.QuestionSkillLinkModel
+        ]
+    ) -> int:
+        """Counts the number of unique question ids.
+
+        Args:
+            question_skill_link_models: list(QuestionSkillLinkModel).
+                List of QuestionSkillLinkModels.
+
+        Returns:
+            int. The number of unique question ids.
+        """
+
+        return len({link.question_id for link in question_skill_link_models})
+
+    @staticmethod
+    def _create_skill_opportunity_model(
+        skill: skill_models.SkillModel,
+        question_skill_links: List[question_models.QuestionSkillLinkModel]
+    ) -> result.Result[
+        opportunity_models.SkillOpportunityModel, Exception
+    ]:
+        """Transforms a skill object and a list of QuestionSkillLink objects
+        into a skill opportunity model.
+
+        Args:
+            skill: skill_models.SkillModel. The skill to create the opportunity
+                for.
+            question_skill_links: list(question_models.QuestionSkillLinkModel).
+                The list of QuestionSkillLinkModel for the given skill.
+
+        Returns:
+            Result[opportunity_models.SkillOpportunityModel, Exception].
+            Result object that contains SkillOpportunityModel when the operation
+            is successful and Exception when an exception occurs.
+        """
+        try:
+            skill_opportunity = opportunity_domain.SkillOpportunity( # type: ignore[no-untyped-call]
+                skill_id=skill.id,
+                skill_description=skill.description,
+                question_count=(
+                    GenerateSkillOpportunityModelJob._count_unique_question_ids(
+                        question_skill_links))
+            )
+            skill_opportunity.validate() # type: ignore[no-untyped-call]
+            skill_opportunity_model = opportunity_models.SkillOpportunityModel(
+                id=skill_opportunity.id,
+                skill_description=skill_opportunity.skill_description,
+                question_count=skill_opportunity.question_count
+            )
+            skill_opportunity_model.update_timestamps()
```

Comment:
You will need to add `with datastore_services.get_ndb_context():`
```suggestion
            with datastore_services.get_ndb_context():
                skill_opportunity_model = opportunity_models.SkillOpportunityModel(
                    id=skill_opportunity.id,
                    skill_description=skill_opportunity.skill_description,
                    question_count=skill_opportunity.question_count
                )
                skill_opportunity_model.update_timestamps()
```

Reply:
Done, thanks!